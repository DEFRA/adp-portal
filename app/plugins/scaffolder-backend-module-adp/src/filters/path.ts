import type { JsonValue } from '@backstage/types';
import pathImpl from 'node:path';

export function path(input: JsonValue, ...segments: JsonValue[]) {
  const baseStr = String(input);
  const schemeMatch = /^\w+:\//.exec(baseStr);
  const [scheme, s1] =
    schemeMatch === null
      ? ['', baseStr]
      : [
          baseStr.slice(0, schemeMatch[0].length),
          baseStr.slice(schemeMatch[0].length),
        ];

  return scheme + pathImpl.join(s1, ...segments.map(String));
}
