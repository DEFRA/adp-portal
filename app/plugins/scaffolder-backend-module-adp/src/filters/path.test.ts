import { path } from './path';

const inputs = [
  ['/my/path', 'abc/'],
  ['/my/path', 'abc'],
  ['/my/path/', 'abc'],
  ['/my/path', '/abc'],
  ['/my/path/', '/abc'],
  ['/my/path', './abc'],
  ['/my/path/', './abc'],
  ['https://test.com', 'abc/'],
  ['https://test.com', 'abc'],
  ['https://test.com/', 'abc'],
  ['https://test.com', '/abc'],
  ['https://test.com/', '/abc'],
  ['https://test.com', './abc'],
  ['https://test.com/', './abc'],
  ['C:/Users/Test', 'abc/'],
  ['C:/Users/Test', 'abc'],
  ['C:/Users/Test/', 'abc'],
  ['C:/Users/Test', '/abc'],
  ['C:/Users/Test/', '/abc'],
  ['C:/Users/Test', './abc'],
  ['C:/Users/Test/', './abc'],
] as const;

interface TestCase {
  readonly input: readonly [string, ...string[]];
  readonly inputJSON: string;
}

describe('path', () => {
  it.each<TestCase>(
    inputs.map(input => ({
      input,
      inputJSON: JSON.stringify(input),
    })),
  )('Should join $inputJSON', ({ input }) => {
    // act
    const actual = path(...input);

    // assert
    expect(actual).toMatchSnapshot();
  });
});
