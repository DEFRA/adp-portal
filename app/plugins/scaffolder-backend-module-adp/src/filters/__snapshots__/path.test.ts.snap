// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`path Should join ["/my/path","./abc"] 1`] = `"/my/path/abc"`;

exports[`path Should join ["/my/path","/abc"] 1`] = `"/my/path/abc"`;

exports[`path Should join ["/my/path","abc"] 1`] = `"/my/path/abc"`;

exports[`path Should join ["/my/path","abc/"] 1`] = `"/my/path/abc/"`;

exports[`path Should join ["/my/path/","./abc"] 1`] = `"/my/path/abc"`;

exports[`path Should join ["/my/path/","/abc"] 1`] = `"/my/path/abc"`;

exports[`path Should join ["/my/path/","abc"] 1`] = `"/my/path/abc"`;

exports[`path Should join ["C:/Users/Test","./abc"] 1`] = `"C:/Users/Test/abc"`;

exports[`path Should join ["C:/Users/Test","/abc"] 1`] = `"C:/Users/Test/abc"`;

exports[`path Should join ["C:/Users/Test","abc"] 1`] = `"C:/Users/Test/abc"`;

exports[`path Should join ["C:/Users/Test","abc/"] 1`] = `"C:/Users/Test/abc/"`;

exports[`path Should join ["C:/Users/Test/","./abc"] 1`] = `"C:/Users/Test/abc"`;

exports[`path Should join ["C:/Users/Test/","/abc"] 1`] = `"C:/Users/Test/abc"`;

exports[`path Should join ["C:/Users/Test/","abc"] 1`] = `"C:/Users/Test/abc"`;

exports[`path Should join ["https://test.com","./abc"] 1`] = `"https://test.com/abc"`;

exports[`path Should join ["https://test.com","/abc"] 1`] = `"https://test.com/abc"`;

exports[`path Should join ["https://test.com","abc"] 1`] = `"https://test.com/abc"`;

exports[`path Should join ["https://test.com","abc/"] 1`] = `"https://test.com/abc/"`;

exports[`path Should join ["https://test.com/","./abc"] 1`] = `"https://test.com/abc"`;

exports[`path Should join ["https://test.com/","/abc"] 1`] = `"https://test.com/abc"`;

exports[`path Should join ["https://test.com/","abc"] 1`] = `"https://test.com/abc"`;
